hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AddMigrationListenerMessageTask.java

```diff
@@ -0,0 +1,153 @@
+import static com.hazelcast.spi.impl.InternalCompletableFuture.newCompletedFuture;
+
+public class AddMigrationListenerMessageTask
+        extends AbstractAddListenerMessageTask<Boolean> {
+
+    public AddMigrationListenerMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected CompletableFuture<UUID> processInternal() {
+        IPartitionService partitionService = getService(getServiceName());
+
+        MigrationListener listener = createMigrationListener();
+
+        if (parameters) {
+            return newCompletedFuture(partitionService.addLocalMigrationListener(listener));
+        }
+
+        return partitionService.addMigrationListenerAsync(listener);
+    }
+
+    private MigrationListener createMigrationListener() {
+        return new MigrationListener() {
+
+                @Override
+                public void migrationStarted(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, MIGRATION_STARTED));
+                }
+
+                @Override
+                public void migrationFinished(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, MIGRATION_FINISHED));
+                }
+
+                @Override
+                public void replicaMigrationCompleted(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+
+                @Override
+                public void replicaMigrationFailed(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+            };
+    }
+
+    private void sendIfAlive(ClientMessage eventMessage) {
+        if (endpoint.isAlive()) {
+            sendClientMessage(MIGRATION_EVENT_TOPIC_ORDER_KEY, eventMessage);
+        }
+    }
+
+    private ClientMessage encodeReplicaMigrationEvent(ReplicaMigrationEvent event) {
+        return ClientAddMigrationListenerCodec.encodeReplicaMigrationEvent(
+                new MigrationStateImpl(
+                        event.getMigrationState().getStartTime(),
+                        event.getMigrationState().getPlannedMigrations(),
+                        event.getMigrationState().getCompletedMigrations(),
+                        event.getMigrationState().getTotalElapsedTime()
+                ),
+                event.getPartitionId(),
+                event.getReplicaIndex(),
+                event.getSource() != null ? event.getSource().getUuid() : null,
+                event.getDestination() != null ? event.getDestination().getUuid() : null,
+                event.isSuccess(),
+                event.getElapsedTime()
+        );
+    }
+
+    private ClientMessage encodeMigrationEvent(MigrationState event, int partitionId) {
+        return ClientAddMigrationListenerCodec.encodeMigrationEvent(
+                new MigrationStateImpl(
```

Comment:
No need to create new `MigrationStateImpl` object. We can pass event.getMigrationState to `encode..` method